src/preprocessing/preprocessing_pass_registry.h

```diff
@@ -29,32 +29,102 @@
 namespace CVC4 {
 namespace preprocessing {
 
+class PreprocessingPassContext;
+
+/**
+ * The PreprocessingPassRegistry keeps track of the available preprocessing
+ * passes and how to create instances of those passes. This class is intended
+ * to be used as a singleton and can be shared between threads (assuming that
```

Comment:
Is this really true? We have passes that maintain pass-specific data that is at the same time specific to the current SmtEngine, e.g., https://github.com/CVC4/CVC4/blob/master/src/preprocessing/passes/non_clausal_simp.h#L50.

Reply:
In that case, you'd create multiple instances of a pass. Each SmtEngine gets its own instance.